drivers/xen/regions.c

```diff
@@ -0,0 +1,192 @@
+#define EXT_BLOCK(node_id, idx)						\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),		\
+		     "Extended region starting from 0 is not supported, please skip page 0"); \
+	BUILD_ASSERT(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region size is not aligned");		\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region address is not aligned");		\
+	BUILD_ASSERT(NOT_OVERLAPPING(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx), \
+				     DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx), \
+				     CONFIG_KERNEL_VM_BASE,		\
+				     CONFIG_KERNEL_VM_SIZE),		\
+		     "Extended region should not overlap virtual ram");	\
+	SYS_MEM_BLOCKS_DEFINE_STATIC_WITH_EXT_BUF(			\
+		region_##idx, XEN_PAGE_SIZE,				\
+		XEN_PFN_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx)),	\
+		(uint8_t *)DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx));
+
+DT_FOREACH_REG(XEN_HYP_NODE, EXT_BLOCK);
+
+#define EXT_REG_POINTER(node_id, idx)					\
+	{								\
+		.block = &region_##idx,					\
+		.phys_addr = DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),	\
+		.size = DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx),		\
+	},
+
+/*
+ * TODO: For now we allocate buffer structures for DT_NUM_REGS, when we
+ * need a range 1...DT_FOREACH_REG. Although region 0 is ignored, it still
+ * takes space for sys_block structures allocation. This requires optimization
+ */
+static struct extended_regions ext_blocks[DT_NUM_REGS(XEN_HYP_NODE)] = {
+	DT_FOREACH_REG(XEN_HYP_NODE, EXT_REG_POINTER)
+};
+
+void *xen_region_get_pages(size_t nr_pages)
+{
+	int ret, i;
+	void *ptr;
+
+	/*
+	 * Locking ranges iteration, although each call sys_mem_blocks_alloc_contiguous
+	 * has it's own syncronisation, to avoid the situation of 2 parralel calls of
+	 * get_page and put_page on full buffers when put_page clears space in block 1
+	 * when get_page has already checked block 1. This will lead us to -ENOMEM while
+	 * block 1 space was freed already
+	 */
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		ret = sys_mem_blocks_alloc_contiguous(ext_blocks[i].block, nr_pages, &ptr);
+		if (ret == 0) {
+			break;
+		} else if (ret && ret != -ENOMEM) {
+			LOG_ERR("Unable to allocate pages, ret = %d", ret);
+			ptr = NULL;
+		}
+	}
+
+	k_mutex_unlock(&ext_regions_lock);
+
+	return ptr;
+}
+
+static int addr_is_in_buffer(uint8_t *ptr, sys_mem_blocks_t *block)
+{
+	/* Compare ptr to buffer address to determine if it was allocated in this block */
+	return (ptr >= block->buffer) &&
+		(ptr < block->buffer + (block->num_blocks << block->blk_sz_shift));
+}
+
+void xen_region_put_pages(void *ptr, size_t nr_pages)
+{
+	int ret, i;
+	sys_mem_blocks_t *block = NULL;
+
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
```

Comment:
(note) you can rewrite loop in the next way and avoid goto at all:
```
k_mutex_lock(&ext_regions_lock, K_FOREVER);
for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
	if (addr_is_in_buffer(ptr, ext_blocks[i].block)) {
			block = ext_blocks[i].block;
			ret = sys_mem_blocks_free_contiguous(block, ptr, nr_pages);
			break;
	}
}
k_mutex_unlock(&ext_regions_lock);
/* print logs if necessary */

```

Reply:
I would say this will not increase code readability. In your approach we don't need `block` variable, but we will `swallow` error then block wasn't found. This can be handled of cause but this will be hard to read IMHO.

Reply:
actually it increases readability because code will become more linear
we still need block variable to print error message, or we need to play with ret variable more

moreover it decreases time under lock

Reply:
I would disagree with that. This makes us 2 problems for my vision: 
1) Block- variable will be only for error checking
2) We make ret dangling until we reach 'if' section, wich will make easily to be overwritten by somebody who will add additional call here.
Asl for the time under lock - I would say it's not so significant time reducing by moving 2 checks outside of the mutex.

Reply:
1. not only for error checking but also we use it as arg for call `sys_mem_blocks_free_contiguous`
2. you forget about possible printing in case of errors, moreover somebody who add extra call should add ret checking I see no problem here

Reply:
1) we don't need it in your case. Just call sys_mem_blocks_free_contiguous(ext_regions[idx].block, ptr, nr_pages);
2) here is an example: 
`k_mutex_lock(&ext_regions_lock, K_FOREVER);
for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
	if (addr_is_in_buffer(ptr, ext_blocks[i].block)) {
			block = ext_blocks[i].block;
			ret = sys_mem_blocks_free_contiguous(block, ptr, nr_pages);
			break;
	}
}
ret = some_call(); // We should check if ret == 0 before doing this call or we would overwrite ret
k_mutex_unlock(&ext_regions_lock);
/* print logs if necessary */`

Reply:
why should we avoid such possibility if it isn't problem as you wrote?

Reply:
> And what if we have to add 20 calls? Check ret before each call?

yeah, and it happens in any case )

Reply:
Sorry. It's not common approach. I can't make this change. If you still thinking it's better - you can make PR to this patch series with your Sign-off.

Reply:
emmm... not common approach, Is yours approach common?
few words from MISRA:
```
Rule 15.1 The goto statement should not be used
Category Advisory
Unconstrained use of goto can lead to programs that are unstructured and extremely difficult to
understand.
In some cases a total ban on goto requires the introduction of flags to ensure correct control flow,
and it is possible that these flags may themselves be less transparent than the goto they replace.
Therefore, if this rule is not followed, the restricted use of goto is allowed where that use follows the
guidance in Rule 15.2 and Rule 15.3.
```
so it is possible to avoid using of goto here, at least now

Reply:
again, if somebody need check ret, he or she will add necessary checks and `goto`,
as for now it is possible simplify code and its readability

>This is Advisory.

it is a common recommendation with explanation why you need to avoid goto if it possible, but it isn't an argument for you

everyday there is a possibility (really small) that meteor will break your house but you haven't done nothing with that
you can't predict all cases and leave redundant and ugly code only because there is some possibility that somebody will make mistake in a future during adding his code and changes will successfully pass review process

or maybe lets always create functions that return some status even if we don't need it, somebody can add some valid status in the future :)

I'm really tired of arguing with you

Reply:
That's not argument for me because I'm still thinking that avoiding goto will make this code ugly and redundand. I've descibed why in https://github.com/xen-troops/zephyr/pull/44#discussion_r1206900962
@lorc  @firscity  what do you think about that? If you say so - I would make this change.

Reply:
Guys, you are surprising me.

@xakep-amatop please refrain from (note) or (suggestion) comments in the future. It is rarely only one correct way to do a thing. In most cases that are multiple valid approaches. If you have some hard concerns about code - provide them, of course. But comments like "There is another approach which I like better"  are counter-productive. I can write dozens of such comments for your patches, but I am not doing this.

As for this particular case - it can be done in both ways. I am suggesting to leave it as is. Personally I like the current approach more  because we have distinct "lookup" and "free" parts. But I am no imposing this varian, I just saying that there is no need to rework it. It is good as it is.

Reply:
so we can easily avoid violation of MISRA rule without any side effect, but we should chose already implemented version because somebody don't want to redo
it's just some kind of double standards

@lorc ok, I got your point

Reply:
@xakep-amatop , MISRA C Rule 15.1 is advisory. Moreover, as you cited

> In some cases a total ban on goto requires the introduction of flags to ensure correct control flow,
and it is possible that these flags may themselves be less transparent than the goto they replace.
Therefore, if this rule is not followed, the restricted use of goto is allowed where that use follows the
guidance in Rule 15.2 and Rule 15.3.

Also, you didn't begin this thread with statement "This implementation violates MISRA C Rule 15.1",  instead you started with vague suggestion "you can do this way". This sparked unhealthy conversation and you brought MISRA only in the middle of it. So, I am repeating this again: if you spotted a problem, standard violation, potential bug - just say about it: "This is wrong because..." or "In situation A, event B may happen, this leads to a problem C". I already saw such comments from you. You are really good in spotting potential problems. But do not tell others how exactly write code. Let author decide.

And by the way, your proposed code violates MISRA Directive 4.7: "[...] Whenever such a mechanism is provided
by a function the calling program shall check for the indication of an error **as soon** as the function
returns."